File: app/src/main/java/com/tomclaw/appsend/main/item/AppItem.java
package com.tomclaw.appsend.main.item;

import android.content.pm.PackageInfo;
import android.os.Parcel;

import com.tomclaw.appsend.net.UpdatesCheckInteractor.AppEntry;

/**
 * Created by Solkin on 11.12.2014.
 */
public class AppItem extends CommonItem {

    private long firstInstallTime;
    private long lastUpdateTime;
    private AppEntry update;

    public AppItem() {
    }

    public AppItem(String label, String packageName, String version, String path, long size,
                   long firstInstallTime, long lastUpdateTime, AppEntry update, PackageInfo packageInfo) {
        super(label, packageName, version, path, size, packageInfo);
        this.firstInstallTime = firstInstallTime;
        this.lastUpdateTime = lastUpdateTime;
        this.update = update;
    }

    private AppItem(Parcel in) {
        super(in);
        firstInstallTime = in.readLong();
        lastUpdateTime = in.readLong();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeLong(firstInstallTime);
        dest.writeLong(lastUpdateTime);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<AppItem> CREATOR = new Creator<AppItem>() {
        @Override
        public AppItem createFromParcel(Parcel in) {
            return new AppItem(in);
        }

        @Override
        public AppItem[] newArray(int size) {
            return new AppItem[size];
        }
    };

    public long getFirstInstallTime() {
        return firstInstallTime;
    }

    public long getLastUpdateTime() {
        return lastUpdateTime;
    }

    public AppEntry getUpdate() {
        return update;
    }
}

File: app/src/main/java/com/tomclaw/appsend/screen/store/adapter/app/AppItem.kt
package com.tomclaw.appsend.screen.store.adapter.app

import android.os.Parcel
import android.os.Parcelable
import com.avito.konveyor.blueprint.Item
import com.tomclaw.appsend.util.readBool
import com.tomclaw.appsend.util.writeBool

class AppItem(
    override val id: Long,
    val appId: String,
    val icon: String?,
    val title: String,
    val version: String,
    val size: String,
    val rating: Float,
    val downloads: Int,
    var isNew: Boolean = false,
    var hasMore: Boolean = false,
    var hasProgress: Boolean = false,
) : Item, Parcelable {

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeLong(id)
        writeString(appId)
        writeString(icon)
        writeString(title)
        writeString(version)
        writeString(size)
        writeFloat(rating)
        writeInt(downloads)
        writeBool(isNew)
        writeBool(hasMore)
        writeBool(hasProgress)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<AppItem> {
        override fun createFromParcel(parcel: Parcel): AppItem {
            val id = parcel.readLong()
            val appId = parcel.readString().orEmpty()
            val icon = parcel.readString()
            val title = parcel.readString().orEmpty()
            val version = parcel.readString().orEmpty()
            val size = parcel.readString().orEmpty()
            val rating = parcel.readFloat()
            val downloads = parcel.readInt()
            val isNew = parcel.readBool()
            val hasMore = parcel.readBool()
            val hasProgress = parcel.readBool()
            return AppItem(
                id,
                appId,
                icon,
                title,
                version,
                size,
                rating,
                downloads,
                isNew,
                hasMore,
                hasProgress
            )
        }

        override fun newArray(size: Int): Array<AppItem?> {
            return arrayOfNulls(size)
        }
    }

}
